feat(stepfunctions-tasks): assign boolean value in DynamoDB from state input (Json path)

michaelwiles opened the pull request:
It is now possible to map a boolean value from state input using a Json path.

function `booleanFromJsonPath` added to `DynamoAttributeValue` which accepts
a Json path

Closes #9007


----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 442a1eebf0bc6ca1f559031a551bf7040b585916
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d1592ff0-e371-44d7-8d6b-01bbf243dd6b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented on packages/@aws-cdk/aws-stepfunctions-tasks/lib/dynamodb/shared-types.ts:
```diff
@@ -232,6 +232,14 @@ export class DynamoAttributeValue {
   }
 
+  /**
+   * maps a boolean from a string - can be a literal or a json path
```
I think this should not support string literals and validate the JsonPath similar to how `mapFromJsonPath` method is laid out.

If it turns out we really need it, we can always expand later. I can't really think of a scenario where a string literal would be needed as all supported languages should be able to use `fromBoolean` instead. It also helps prevent user errors from providing strings like "foo" when only "true" and "false" are valid.

You could extract that validation logic from `mapFromJsonPath` into the `.private` folder

michaelwiles replied:
I agree - makes a lot of sense

michaelwiles replied:
Are you saying it should actually take a JsonPath? or is confirming it starts with $ adequate in this regard?

shivlaks replied:
yep, validating that it starts with `$` is what I meant which is roughly the same as accepting only Json paths

shivlaks commented on packages/@aws-cdk/aws-stepfunctions-tasks/lib/dynamodb/shared-types.ts:
```diff
@@ -232,6 +232,14 @@ export class DynamoAttributeValue {
+   * maps a boolean from a string - can be a literal or a json path
+   * @param value the string to map
+   */
+  public static booleanFromJsonPath(value: string) {
```
there should be a unit test added for this method (happy path && Json path error throws)

michaelwiles replied:
good

michaelwiles commented:
I've addressed those issues.

@shivlaks I've added you as a committer on that repo.

It'd probably be easier for you to make changes you need. I'm not sure if you want for example the validate method to throw an exception or not and it'd be easier if you just made any changes yourself.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ba50171f5c0d275d6c0d5886a4b2501d57124f07
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=889b5a79-6097-4a98-b4c6-3061ad729545%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 4490f5bf819efd14c11e2a4e4bf4fe267f12165c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=afb35568-1c52-43e7-a255-fa99f5da28ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented:
> I've addressed those issues.
> 
> @shivlaks I've added you as a committer on that repo.
> 
> It'd probably be easier for you to make changes you need. I'm not sure if you want for example the validate method to throw an exception or not and it'd be easier if you just made any changes yourself.

thanks! I think you've addressed my feedback. might update the doc string for the method slightly, but other than that this is fantastic!

shivlaks commented on packages/@aws-cdk/aws-stepfunctions-tasks/lib/dynamodb/shared-types.ts:
```diff
@@ -232,6 +230,15 @@ export class DynamoAttributeValue {
     return new DynamoAttributeValue({ BOOL: value });
   }
 
+  /**
+   * maps a boolean from a string - can be a literal or a json path
+   * @param value the string to map
+   */
```
```suggestion
  /**
   * Sets an attribute of type Boolean from state input through Json path.
   * For example:  "BOOL": true
   *
   * @param value Json path that specifies state input to be used
   */
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 845b9f26f0d17e982e6b7ffd348510d741526871
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d2a90b23-094e-4f59-b2ca-a473f27327ca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

shivlaks commented:
@michaelwiles you might need to give maintainers permissions so mergify can update and merge the PR.

here is the [GitHub documentation](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on how you can provide maintainers permissions to the PR

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 351a6bbc68af11d83e9b77e06f3f5cfc67a6f209
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=14e8fc49-6205-43e0-8df6-64dad555e558%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

RomainMuller commented:
@Mergifyio refresh

RomainMuller commented:
This PR will not auto-merge because it's behind `master` and we don't have permissions to update the PR branch. @michaelwiles, would you be able to edit the PR to allow us to make changes there (it's a per-PR setting)? Otherwise, we'll just have to manually merge this :)

michaelwiles commented:
> @michaelwiles you might need to give maintainers permissions so mergify can update and merge the PR.

Yeah... being new to this I made the mistake of creating the pull request repo in an organisation account and that "allow edits from maintainers" check box is thus not available.

michaelwiles commented:
Yeah, like I said I'm new to this so I naively used an organisation repo
and with an organisation repo you can't allow changes to be pushed from the
pull request
<https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork>
.

So unless I create a whole new pull request from a user repository (I don't
think that would be a good idea as I'd assume we'd lose the pull request
history).

I am sorry, will know for next time (let's hope there is a next time).

On Fri, 17 Jul 2020 at 10:21, Romain Marcadier <notifications@github.com>
wrote:

> This PR will not auto-merge because it's behind master and we don't have
> permissions to update the PR branch. @michaelwiles
> <https://github.com/michaelwiles>, would you be able to edit the PR to
> allow us to make changes there (it's a per-PR setting)? Otherwise, we'll
> just have to manually merge this :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aws/aws-cdk/pull/9088#issuecomment-659954065>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAODX5EDYN7HXBZF7L2PDEDR4ACZBANCNFSM4O2ZMQPA>
> .
>


-- 
see my blog:
http://analysis102.blogspot.com
http://audiblethoughts.blogspot.com
http://outsideofficehours.blogspot.com

shivlaks commented:
@michaelwiles - it's not a big deal if you want to create another PR. You can always mention that in the description that it supersedes this one so we have a breadcrumb back to the PR history.

Alternatively, I could merge this although I think that's the least desirable option as it might not pick up an in-flight change that might conflict. Let me know if it's possible to re-publish and we'll go from there to work with the best option in front of us.

We really appreciate the contribution and would love to see more!

shivlaks commented:
@michaelwiles - I merged it this time since it's less painful.

I'll add a note re: pull requests from organization repos since I was not aware that they could not delegate permissions to maintainers. At least it's on your radar for the next time! thanks a bunch  🎉
